lake-primitives/src/types/receipts.rs

```diff
@@ -148,113 +151,687 @@ impl Action {
+            Err("Only `ReceiptEnumView::Action` can be converted into Vec<ActionKind>")
+        }
+    }
+
+    // Tries to convert a near_primitives::delegate_action::DelegateAction into a vector of ActionKind.
+    pub fn try_from_delegate_action(
+        delegate_action: &near_primitives::delegate_action::DelegateAction,
+        metadata: ActionMetadata,
+    ) -> Result<Vec<Self>, &'static str> {
+        let mut actions = Vec::with_capacity(delegate_action.actions.len());
+
+        for action in delegate_action.clone().actions {
+            let action_kind = match views::ActionView::from(
+                <near_primitives::delegate_action::NonDelegateAction as Into<
+                    near_primitives::transaction::Action,
+                >>::into(action),
+            ) {
+                views::ActionView::CreateAccount => ActionKind::CreateAccount(CreateAccount {
+                    metadata: metadata.clone(),
+                }),
+                views::ActionView::DeployContract { code } => {
+                    ActionKind::DeployContract(DeployContract {
+                        metadata: metadata.clone(),
+                        code,
+                    })
+                }
+                views::ActionView::FunctionCall {
+                    method_name,
+                    args,
+                    gas,
+                    deposit,
+                } => ActionKind::FunctionCall(FunctionCall {
+                    metadata: metadata.clone(),
+                    method_name,
+                    args,
+                    gas,
+                    deposit,
+                }),
+                views::ActionView::Transfer { deposit } => ActionKind::Transfer(Transfer {
+                    metadata: metadata.clone(),
+                    deposit,
+                }),
+                views::ActionView::Stake { stake, public_key } => ActionKind::Stake(Stake {
+                    metadata: metadata.clone(),
+                    stake,
+                    public_key,
+                }),
+                views::ActionView::AddKey {
+                    public_key,
+                    access_key,
+                } => ActionKind::AddKey(AddKey {
+                    metadata: metadata.clone(),
+                    public_key,
+                    access_key,
+                }),
+                views::ActionView::DeleteKey { public_key } => ActionKind::DeleteKey(DeleteKey {
+                    metadata: metadata.clone(),
+                    public_key,
+                }),
+                views::ActionView::DeleteAccount { beneficiary_id } => {
+                    ActionKind::DeleteAccount(DeleteAccount {
+                        metadata: metadata.clone(),
+                        beneficiary_id,
+                    })
+                }
+                views::ActionView::Delegate { .. } => {
+                    return Err("Cannot delegate DelegateAction variant")
+                }
```

Comment:
I don't have a solution, but I would prefer to not even be able to reach this state, so it should be prohibited on the types level, and then we don't need to use Result here, as the method would always succeed.

Reply:
On the `nearcore` side it is achieved by introducing `NonDelegateAction` structure that is a wrapper over the `Action` and a `TryFrom` implemented that returns an error. I am not a fan of either that solution or mine. So I don't have a better solution either :(